### PR TITLE
feat(Auth): HostedUI Start / Signout

### PR DESCRIPTION
**Pull request by tylerjroach**
- [ ] PR title and description conform to [Pull Request](https://github.com/aws-amplify/amplify-android/blob/main/CONTRIBUTING.md#pull-request-guidelines) guidelines.

*Issue #, if available:*

*Description of changes:*
The beginnings of HostedUI feature supporting web sign in and social sign in.
Added Sign Out

Missing:
* Federation
* Tests (Waiting on more clarity around federation and iOS alignment on error states)
* Refresh Token

*How did you test these changes?*
Manual happy path testing. This PR is not feature complete hosted ui

- [ ] Added Unit Tests
- [ ] Added Integration Tests

*Documentation update required?*
- [ ] No
- [ ] Yes (Please include a PR link for the documentation update)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/actions/HostedUICognitoActions.kt [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import com.amplifyframework.statemachine.codegen.events.HostedUIEvent
+import java.util.Date
+
+object HostedUICognitoActions : HostedUIActions {
+
+    override fun showHostedUI(event: HostedUIEvent.EventType.ShowHostedUI) =
+        Action<AuthEnvironment>("InitHostedUIAuth") { id, dispatcher ->
+            logger?.verbose("$id Starting execution")
+            try {
+                if (hostedUIClient == null) throw Exception() // TODO: More detailed exception
```
Comment (gpanshu) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/actions/HostedUICognitoActions.kt:34]: We have exceptions inside AuthException which can be thrown here. Worst case just throw a generic Auth Exception?

Comment (tylerjroach) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/actions/HostedUICognitoActions.kt:34]: Yeah some of these can be generic. I wanted to do an audit of ours/iOS together to match any error messaging that could be returned, so I threw todos anywhere I hadn't reviewed yet.

File: aws-auth-cognito/src/main/java/com/amplifyframework/statemachine/codegen/data/AuthConfiguration.kt [hunk 2]
```diff
@@ -20,77 +20,36 @@ import org.json.JSONObject
...
+                        IdentityPoolConfiguration.fromJson(it).build()
+                    },
+                oauth = pluginJson.optJSONObject("Auth")
+                    ?.optJSONObject(configName)
+                    ?.optJSONObject("OAuth")?.let {
+                        OauthConfiguration.fromJson(it)
                     }
-            }
+            )
         }
```
Comment (tylerjroach) [inline aws-auth-cognito/src/main/java/com/amplifyframework/statemachine/codegen/data/AuthConfiguration.kt:53]: add constants back

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/actions/HostedUICognitoActions.kt (hunk 1 above)
Comment (tylerjroach) [reply aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/actions/HostedUICognitoActions.kt:34]: Will be following up with a PR updating exception messaging.

**Review by gpanshu — APPROVED** [review]

**Review by sktimalsina — APPROVED** [review]

**Pull request merged**